logsapi/functionlogs.go

```diff
@@ -0,0 +1,99 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package logsapi
+
+import (
+	"github.com/elastic/apm-aws-lambda/apmproxy"
+	"github.com/elastic/apm-aws-lambda/extension"
+	"go.elastic.co/apm/v2/model"
+	"go.elastic.co/fastjson"
+)
+
+type LogContainer struct {
+	Log *logLine
+}
+
+type logLine struct {
+	Timestamp model.Time
+	Message   string
+	FAAS      *model.FAAS
+}
+
+func (l *logLine) MarshalFastJSON(w *fastjson.Writer) error {
```

Comment:
Just an idea that could reduce allocations: instead of converting `extension.NextEventResponse` and `LogEvent` into a `LogContainer`, you could pass in `extension.NextEventResponse` and `LogEvent` directly to the `MarshalFastJSON` method and avoid creating these temporary serialization objects.

However, if you feel like the code is cleaner this way and if this isn't contributing a lot to overall allocations, feel free to ignore.

Reply:
I ran some benchmarks on the idea; below are the results:

```
# With `LogContainer` and `logLine`
BenchmarkProcessFunctionLog-10    	 1758667	       684.3 ns/op	    1264 B/op	       7 allocs/op

# Without `LogContainer` and `logLine`
BenchmarkProcessFunctionLog-10    	 1808710	       668.9 ns/op	    1184 B/op	       6 allocs/op
```

Based on the results I am a bit more inclined toward keeping the objects since they will allow for easier extension in case more fields need to be added.

WDYT given the benchmark results?

<details>
<summary>(Code diff that produced the 2nd benchmark)</summary>

```diff
diff --cc logsapi/functionlogs.go
index 984b15f,3213428..0000000
--- a/logsapi/functionlogs.go
+++ b/logsapi/functionlogs.go
@@@ -24,28 -24,61 +24,23 @@@ import 
  	"go.elastic.co/fastjson"
  )
  
- type logContainer struct {
 -type LogContainer struct {
--	Log *logLine
--}
--
--type logLine struct {
--	Timestamp model.Time
--	Message   string
--	FAAS      *model.FAAS
--}
--
--func (l *logLine) MarshalFastJSON(w *fastjson.Writer) error {
 -	var firstErr error
 -	w.RawByte('{')
 -	w.RawString("\"message\":")
 -	w.String(l.Message)
 -	w.RawString(",\"@timestamp\":")
 -	if err := l.Timestamp.MarshalFastJSON(w); err != nil && firstErr == nil {
 -		firstErr = err
 -	}
 -	if l.FAAS != nil {
 -		w.RawString(",\"faas\":")
 -		if err := l.FAAS.MarshalFastJSON(w); err != nil && firstErr == nil {
 -			firstErr = err
 -		}
 -	}
 -	w.RawByte('}')
 -	return firstErr
 -}
 -
 -func (lc LogContainer) MarshalFastJSON(json *fastjson.Writer) error {
 -	json.RawString(`{"log":`)
 -	if err := lc.Log.MarshalFastJSON(json); err != nil {
 -		return err
 -	}
 -	json.RawString(`}`)
 -	return nil
 -}
 -
+ func marshalFastJSON(w *fastjson.Writer, fd *extension.NextEventResponse, log LogEvent) error {
 -	var firstErr error // 4
 +	var firstErr error
+ 	w.RawString(`{"log":`)
  	w.RawByte('{')
  	w.RawString("\"message\":")
- 	w.String(l.Message)
+ 	w.String(log.StringRecord)
  	w.RawString(",\"@timestamp\":")
- 	if err := l.Timestamp.MarshalFastJSON(w); err != nil && firstErr == nil {
+ 	if err := model.Time(log.Time).MarshalFastJSON(w); err != nil && firstErr == nil {
  		firstErr = err
  	}
- 	if l.FAAS != nil {
+ 	if fd != nil {
  		w.RawString(",\"faas\":")
- 		if err := l.FAAS.MarshalFastJSON(w); err != nil && firstErr == nil {
+ 		faas := model.FAAS{
+ 			Execution: fd.RequestID,
+ 			ID:        fd.InvokedFunctionArn,
+ 		}
+ 		if err := faas.MarshalFastJSON(w); err != nil && firstErr == nil {
  			firstErr = err
  		}
  	}
@@@ -69,23 -94,8 +56,8 @@@ func ProcessFunctionLog
  	functionData *extension.NextEventResponse,
  	log LogEvent,
  ) (apmproxy.AgentData, error) {
- 	lc := logContainer{
- 		Log: &logLine{
- 			Timestamp: model.Time(log.Time),
- 			Message:   log.StringRecord,
- 		},
- 	}
- 
- 	if functionData != nil {
- 		// FaaS Fields
- 		lc.Log.FAAS = &model.FAAS{
- 			Execution: functionData.RequestID,
- 			ID:        functionData.InvokedFunctionArn,
- 		}
- 	}
- 
 -	var jsonWriter fastjson.Writer // 1
 +	var jsonWriter fastjson.Writer
- 	if err := lc.MarshalFastJSON(&jsonWriter); err != nil {
+ 	if err := marshalFastJSON(&jsonWriter, functionData, log); err != nil {
  		return apmproxy.AgentData{}, err
  	}
  
diff --git a/logsapi/functionlogs_test.go b/logsapi/functionlogs_test.go
index 64d4312..5aed6a6 100644
--- a/logsapi/functionlogs_test.go
+++ b/logsapi/functionlogs_test.go
@@ -68,3 +68,23 @@ func TestProcessFunctionLog(t *testing.T) {
 		)
 	})
 }
+
+func BenchmarkProcessFunctionLog(b *testing.B) {
+	mc := &apmproxy.MetadataContainer{
+		Metadata: []byte(`{"metadata":{"service":{"agent":{"name":"apm-lambda-extension","version":"1.1.0"},"framework":{"name":"AWS Lambda","version":""},"language":{"name":"python","version":"3.9.8"},"runtime":{"name":"","version":""},"node":{}},"user":{},"process":{"pid":0},"system":{"container":{"id":""},"kubernetes":{"node":{},"pod":{}}},"cloud":{"provider":"","instance":{},"machine":{},"account":{},"project":{},"service":{}}}}`),
+	}
+	event := LogEvent{
+		Time:         time.Date(2022, 11, 12, 0, 0, 0, 0, time.UTC),
+		Type:         FunctionLog,
+		StringRecord: "ERROR encountered. Stack trace:\nmy-function (line 10)\n",
+	}
+	nextEventResp := &extension.NextEventResponse{
+		RequestID:          "8476a536-e9f4-11e8-9739-2dfe598c3fcd",
+		InvokedFunctionArn: "arn:aws:lambda:us-east-2:123456789012:function:custom-runtime",
+	}
+
+	for n := 0; n < b.N; n++ {
+		_, err := ProcessFunctionLog(mc, nextEventResp, event)
+		require.NoError(b, err)
+	}
+}
```
</details>

Reply:
Seems to not make a huge difference allocation-wise.

> Based on the results I am a bit more inclined toward keeping the objects since they will allow for easier extension in case more fields need to be added.

SGTM